sky/spot/controller.py

```diff
@@ -167,40 +166,82 @@ def start(self):
+    # the controller is not serializable.
+    spot_controller = SpotController(job_id, task_yaml, retry_until_up)
+    spot_controller.run()
+
+
+def _handle_signal(job_id):
+    """Handle the signal if the user sent it."""
+    signal_file = pathlib.Path(spot_utils.SIGNAL_FILE_PREFIX.format(job_id))
+    signal = None
+    if signal_file.exists():
+        # Filelock is needed to prevent race condition with concurrent
+        # signal writing.
+        # TODO(mraheja): remove pylint disabling when filelock version
+        # updated
+        # pylint: disable=abstract-class-instantiated
+        with filelock.FileLock(str(signal_file) + '.lock'):
+            with signal_file.open(mode='r') as f:
+                signal = f.read().strip()
+                try:
                     signal = spot_utils.UserSignal(signal)
-                # Remove the signal file, after reading the signal.
-                signal_file.unlink()
-        if signal is None:
-            return
-        if signal == spot_utils.UserSignal.CANCEL:
-            raise exceptions.SpotUserCancelledError(
-                f'User sent {signal.value} signal.')
+                except ValueError:
+                    logger.warning(
+                        f'Unknown signal received: {signal}. Ignoring.')
+                    signal = None
+            # Remove the signal file, after reading the signal.
+            signal_file.unlink()
+    if signal is None:
+        # None or empty string.
+        return
+    assert signal == spot_utils.UserSignal.CANCEL, (
+        f'Only cancel signal is supported, but {signal} got.')
+    raise exceptions.SpotUserCancelledError(f'User sent {signal.value} signal.')
+
 
-        raise RuntimeError(f'Unknown SkyPilot signal received: {signal.value}.')
+def start(job_id, task_yaml, retry_until_up):
+    """Start the controller."""
+    controller_task = None
+    try:
+        _handle_signal(job_id)
+        controller_task = _run_controller.remote(job_id, task_yaml,
+                                                 retry_until_up)
+        # Signal can interrupt the underlying controller process.
+        ready, _ = ray.wait([controller_task], timeout=0)
+        while not ready:
+            _handle_signal(job_id)
+            ready, _ = ray.wait([controller_task], timeout=1)
+    except exceptions.SpotUserCancelledError:
+        logger.info(f'Cancelling spot job {job_id}...')
+        try:
+            if controller_task is not None:
+                # This will raise KeyboardInterrupt in the task.
+                ray.cancel(controller_task)
+                ray.get(controller_task)
+        except ray.exceptions.RayTaskError:
+            # When the controller task is cancelled, it will raise
+            # ray.exceptions.RayTaskError, which can be ignored,
+            # since the SpotUserCancelledError will be raised and
+            # handled later.
+            pass
 
 
 if __name__ == '__main__':
+    ray.init()
```

Comment:
Can we move this `ray.init()` call to inside `start()`? That way any external module can also directly use the spot controller by simply calling `controller.start()`.

Reply:
Good point! Moved. Thanks!